packages/core/graph/src/AdjacencyList.js

```diff
@@ -0,0 +1,1210 @@
+      bucket.add(key);
+      buckets.set(hash, bucket);
+    }
+
+    let maxCollisions = 0;
+    let collisions = 0;
+    let distribution = 0;
+
+    for (let bucket of buckets.values()) {
+      maxCollisions = Math.max(maxCollisions, bucket.size - 1);
+      collisions += bucket.size - 1;
+      distribution += (bucket.size * (bucket.size + 1)) / 2;
+    }
+
+    let uniformity =
+      distribution /
+      ((this.#edges.count / (2 * this.#edges.capacity)) *
+        (this.#edges.count + 2 * this.#edges.capacity - 1));
+
+    return {
+      nodes: fromNodeId(this.#nodes.nextId),
+      nodeEdgeTypes: this.#nodes.count,
+      nodeCapacity: this.#nodes.capacity,
+      nodeLoad: `${Math.round(this.#nodes.load * 100)}%`,
+      nodeBufferSize: this.#nodes.bufferSize,
+
+      edges: this.#edges.count,
+      deleted: this.#edges.deletes,
+      edgeCapacity: this.#edges.capacity,
+      edgeLoad: `${Math.round(this.#edges.load * 100)}%`,
+      edgeLoadWithDeletes: `${Math.round(
+        this.#edges.getLoad(this.#edges.count + this.#edges.deletes) * 100,
+      )}%`,
+      edgeBufferSize: this.#edges.bufferSize,
+
+      collisions,
+      maxCollisions,
+      avgCollisions: Math.round((collisions / buckets.size) * 100) / 100 || 0,
+      uniformity: Math.round(uniformity * 100) / 100 || 0,
+    };
+  }
+
+  /**
+   * Resize the internal nodes array.
+   *
+   * This is used in `addNode` when the `numNodes` meets or exceeds
+   * the allocated size of the `nodes` array.
+   */
+  resizeNodes(size: number) {
+    let nodes = this.#nodes;
+    // Allocate the required space for a `nodes` map of the given `size`.
+    this.#nodes = new NodeTypeMap(size);
+    // Copy the existing nodes into the new array.
+    this.#nodes.set(nodes.data);
+  }
+
+  /**
+   * Resize the internal edges array.
+   *
+   * This is used in `addEdge` when the `numEdges` meets or exceeds
+   * the allocated size of the `edges` array.
+   */
+  resizeEdges(size: number) {
+    // Allocate the required space for new `nodes` and `edges` maps.
+    let copy = new AdjacencyList({
+      nodeCapacity: this.#nodes.capacity,
+      edgeCapacity: size,
+    });
+
+    // Copy the existing edges into the new array.
+    copy.#nodes.nextId = this.#nodes.nextId;
+    this.#edges.forEach(
+      edge =>
+        void copy.addEdge(
+          this.#edges.from(edge),
+          this.#edges.to(edge),
+          this.#edges.typeOf(edge),
+        ),
+    );
+
+    // We expect to preserve the same number of edges.
+    assert(
+      this.#edges.count === copy.#edges.count,
+      `Edge mismatch! ${this.#edges.count} does not match ${
+        copy.#edges.count
+      }.`,
+    );
+
+    // Finally, copy the new data arrays over to this graph.
+    this.#nodes = copy.#nodes;
+    this.#edges = copy.#edges;
+  }
+
+  /**
+   * Adds a node to the graph.
+   *
+   * Returns the id of the added node.
+   */
+  addNode(): NodeId {
+    let id = this.#nodes.getId();
```

Comment:
Note that we don't _actually_ add an entry to the node map until it is connected  to another node id in `addEdge`.